### PR TITLE
Add In-Memory Support for Set Operations (e.g. Union, Intersect...)

### LINKED ISSUE #3638: Query: Support in-memory processing of INTERSECT,CONCAT,EXCEPT,UNION
Hi ,

I am using EF7BETA6 and trying to run simple set operators and getting an exception.
I think the issue is similar to "Problem with Union Operator #3459 " 

Sample Program : 

 var query = context.ORDERS.Where(o => o.SHIPCOUNTRY == "Mexico").Select(o => o.Customer).
                Except(context.ORDERS.Where(o => o.SHIPCOUNTRY == "Canada").Select(o => o.Customer));

            foreach (var val in query) // getting exception at this point.

-----
    var mexico = context.ORDERSS.Where(o => o.SHIPCOUNTRY == "Mexico").Select(o => o);
            var canada = context.ORDERSS.Where(o => o.SHIPCOUNTRY == "Canada").Select(o => o);

            var query = mexico.Concat(canada);
 foreach (var val in query) // getting exception at this point.

----
Similarly for Intersect and Union.

Stack trace :
 at Microsoft.Data.Entity.Query.ResultOperatorHandler.HandleResultOperator(EntityQueryModelVisitor entityQueryModelVisitor, ResultOperatorBase resultOperator, QueryModel queryModel) in C:\Users\praskarn\Desktop\EF7\EntityFramework-7.0.0-beta6\src\EntityFramework.Core\Query\ResultOperatorHandler.cs:line 58
   at Microsoft.Data.Entity.Query.RelationalResultOperatorHandler.HandlerContext.get_EvalOnClient() in C:\Users\praskarn\Desktop\EF7\EntityFramework-7.0.0-beta6\src\EntityFramework.Relational\Query\RelationalResultOperatorHandler.cs:line 59
   at Microsoft.Data.Entity.Query.RelationalResultOperatorHandler.HandleResultOperator(EntityQueryModelVisitor entityQueryModelVisitor, ResultOperatorBase resultOperator, QueryModel queryModel) in C:\Users\praskarn\Desktop\EF7\EntityFramework-7.0.0-beta6\src\EntityFramework.Relational\Query\RelationalResultOperatorHandler.cs:line 117
   at Microsoft.Data.Entity.Query.EntityQueryModelVisitor.VisitResultOperator(ResultOperatorBase resultOperator, QueryModel queryModel, Int32 index) in 

Please let me know if the issue is fixed in new beta releases.

Thanks and Regards,
Prashanth.

### PR DESCRIPTION
**Pull request by smitpatel**
Resolves #3638

### CODE REVIEW

**Comment by smitpatel** [comment]
@anpete - Can you give some feedback on this about fixing the issue with QuerySourceTracing?

File: src/Microsoft.EntityFrameworkCore/Query/ExpressionVisitors/Internal/RequiresMaterializationExpressionVisitor.cs [hunk 1]
```diff
@@ -34,6 +35,7 @@ public virtual ISet<IQuerySource> FindQuerySourcesRequiringMaterialization([NotN
         {
             _queryModel = queryModel;
             _parentSelector = queryModel.SelectClause.Selector;
+            _parentQueryModel = queryModel;
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore/Query/ExpressionVisitors/Internal/RequiresMaterializationExpressionVisitor.cs]: It looks like we can just use _queryModel to store the current query. I think it means we can remove _parentSelector, too.

File: test/Microsoft.EntityFrameworkCore.FunctionalTests/QueryTestBase.cs [hunk 2]
```diff
@@ -4734,6 +4734,14 @@ public virtual void Throws_on_concurrent_query_first()
             }
         }
 
+        [ConditionalFact]
```
Comment (anpete) [inline test/Microsoft.EntityFrameworkCore.FunctionalTests/QueryTestBase.cs]: Add a test with > 1 Intersects.
Add a test with a non-entity result type.

File: src/Microsoft.EntityFrameworkCore/Query/ExpressionVisitors/Internal/RequiresMaterializationExpressionVisitor.cs [hunk 3]
```diff
@@ -165,6 +170,20 @@ var resultQuerySource
                 {
                     _querySources[querySourceReferenceExpression.ReferencedQuerySource]--;
                 }
+
+                foreach (var intersectOperator in _parentQueryModel.ResultOperators.OfType<IntersectResultOperator>())
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore/Query/ExpressionVisitors/Internal/RequiresMaterializationExpressionVisitor.cs]: Looks good.

File: src/Microsoft.EntityFrameworkCore/Query/ResultOperatorHandler.cs [hunk 4]
```diff
@@ -111,6 +115,20 @@ var resultItemTypeInfo
                 entityQueryModelVisitor.Expression);
         }
 
+        private static Expression HandleConcat(
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore/Query/ResultOperatorHandler.cs]: Consider creating helper method: HandleSetOperation, parameterized on operator method.

**Comment by anpete** [comment]
:shipit:

File: src/Microsoft.EntityFrameworkCore/Query/ResultOperatorHandler.cs (hunk 4 above)
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore/Query/ResultOperatorHandler.cs]: Done

**Comment by smitpatel** [comment]
Updated with `AsyncEnumerable` implementation of `Intersect` & `Except`.

**Comment by smitpatel** [comment]
blocked on #4905

**Comment by smitpatel** [comment]
@anpete - updated on latest changes.

**Comment by smitpatel** [comment]
@anpete - ping for review.

**Pull request merged**